aioxmpp/stream.py

```diff
@@ -927,31 +927,46 @@ def _destroy_stream_state(self, exc):
             self.on_stream_destroyed(exc)
             self._established = False
 
-    def _iq_request_coro_done(self, request, task):
+    def _send_iq_reply(self, request, result):
+        if isinstance(result, errors.XMPPError):
+            response = request.make_reply(type_=structs.IQType.ERROR)
+            response.error = stanza.Error.from_exception(result)
+        else:
+            response = request.make_reply(type_=structs.IQType.RESULT)
+            response.payload = result
+        self._enqueue(response)
+
+    def _iq_request_coro_done_remove_task(self, task):
+        self._iq_request_tasks.remove(task)
+
+    def _iq_request_coro_done_send_reply(self, request, task):
         """
         Called when an IQ request handler coroutine returns. `request` holds
         the IQ request which triggered the excecution of the coroutine and
         `task` is the :class:`asyncio.Task` which tracks the running coroutine.
 
         Compose a response and send that response.
         """
-        self._iq_request_tasks.remove(task)
         try:
             payload = task.result()
         except errors.XMPPError as err:
-            response = request.make_reply(type_=structs.IQType.ERROR)
-            response.error = stanza.Error.from_exception(err)
+            self._send_iq_reply(request, err)
         except Exception:
             response = request.make_reply(type_=structs.IQType.ERROR)
             response.error = stanza.Error(
                 condition=errors.ErrorCondition.UNDEFINED_CONDITION,
                 type_=structs.ErrorType.CANCEL,
             )
+            self._enqueue(response)
             self._logger.exception("IQ request coroutine failed")
         else:
-            response = request.make_reply(type_=structs.IQType.RESULT)
-            response.payload = payload
-        self._enqueue(response)
+            self._send_iq_reply(request, payload)
+
+    def _iq_request_coro_done_check(self, task):
+        try:
+            task.result()
+        except Exception:
+            self._logger.exception("IQ request coroutine failed")
```

Comment:
Raising an XMPP error is not "failure" per se. Just a proper way to signal an error.

Reply:
It is an error if we already sent a reply and let the XMPP error bubble up (and only then will this callback be in place).

Reply:
Ah, I got the flow now, thanks!